src/ameba/rule/lint/unused_pseudo_method_call.cr

```diff
@@ -0,0 +1,78 @@
+module Ameba::Rule::Lint
+  # A rule that disallows unused pseudo method calls (is_a?, sizeof, etc).
+  #
+  # For example, these are considered invalid:
+  #
+  # ```
+  # pointerof(1234_f32)
+  #
+  # method_call.as(Int32)
+  #
+  # def method
+  #   if guard?
+  #     !!valid?
+  #   end
+  #
+  #   true
+  # end
+  # ```
+  #
+  # And these are considered valid:
+  #
+  # ```
+  # a : pointerof(1234_f32)
+  #
+  # var = method_call.as(Int32)
+  #
+  # def method
+  #   if guard?
+  #     return !!valid?
+  #   end
+  #
+  #   true
+  # end
+  # ```
+  #
+  # YAML configuration example:
+  #
+  # ```
+  # Lint/UnusedPseudoMethodCall:
+  #   Enabled: true
+  # ```
+  class UnusedPseudoMethodCall < Base
+    properties do
+      since_version "1.7.0"
+      description "Disallows unused pseudo-method calls"
+    end
+
+    MSG = "Pseudo-method call is not used"
+
+    def test(source : Source)
+      AST::ImplicitReturnVisitor.new(self, source)
+    end
+
+    def test(source, node : Crystal::Not, in_macro : Bool)
+      return unless location = node.location
+
+      position = source.pos(location)
+
+      if source.code[position] == '!'
+        issue_for location, location, MSG
+      else
+        end_location = node.end_location || location
+
+        issue_for end_location, end_location, MSG
+      end
+    end
+
+    def test(
+      source,
+      node : Crystal::PointerOf | Crystal::SizeOf | Crystal::InstanceSizeOf |
+             Crystal::AlignOf | Crystal::InstanceAlignOf | Crystal::OffsetOf |
+             Crystal::IsA | Crystal::NilableCast | Crystal::RespondsTo,
+      in_macro : Bool,
+    )
```

Comment:
This is not needed anymore since the changed message now IMO clarifies the usage.

```suggestion
    def test(
      source,
      node : Crystal::PointerOf | Crystal::SizeOf | Crystal::InstanceSizeOf |
             Crystal::AlignOf | Crystal::InstanceAlignOf | Crystal::OffsetOf |
             Crystal::IsA | Crystal::NilableCast | Crystal::RespondsTo | Crystal::Not,
      in_macro : Bool,
    )
```